File: Kakao/인턴_2020_키패드 누르기.cpp
// 출제의도: "구현, 맨허튼 거리"

#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int manhattan_dist(int finger, int target) {
    return abs((finger-1)/3-(target-1)/3) + abs((finger-1)%3-(target-1)%3);
}

string move_finger_closer(int dist_left, int dist_right, string hand) {
    if(dist_left<dist_right)    return "L";
    else if(dist_left>dist_right)   return "R";
    else {
        if(hand=="left") return  "L";
        else    return "R";
    }
}

string solution(vector<int> numbers, string hand) {
    string answer = "";
    int left=10, right=12;
    for(int i=0;i<numbers.size();i++) {
        if(!numbers[i])    numbers[i]=11;
        
        switch(numbers[i]%3) {
            case 0:
                right = numbers[i];
                answer += "R";
                break;
                
            case 1:
                left = numbers[i];
                answer += "L";
                break;
                
            case 2:
                int dist_left = manhattan_dist(left,numbers[i]);
                int dist_right = manhattan_dist(right,numbers[i]);
                answer += move_finger_closer(dist_left, dist_right, hand);
                if(answer.back()=='L') left = numbers[i];
                else    right = numbers[i];
                break;
        }
    }
    
    return answer;
}

File: Kakao/인턴_2020_키패드 누르기.swift
// 출제의도: "구현", Time Complexity: O(n), Space Complexity: O(n). (n: numbers.size)

import Foundation

struct Point {
    var x: Int
    var y: Int
    
    init(_ x: Int,_ y: Int) {
        self.x = x
        self.y = y
    }
    
    func manhattanDistance(_ target: Point) -> Int {
        return abs(self.x - target.x) + abs(self.y - target.y)
    }
}

var button = [1: Point(0,0), 2: Point(1,0), 3: Point(2,0),
              4: Point(0,1), 5: Point(1,1), 6: Point(2,1),
              7: Point(0,2), 8: Point(1,2), 9: Point(2,2),
                             0: Point(1,3)]

func solution(_ numbers:[Int], _ param1:String) -> String {
    var hand = param1 == "left" ? "L" : "R"
    var leftThumb = Point(0,3)
    var rightThumb = Point(2,3)
    var ans = ""
    
    for number in numbers {
    switch number {
            case 1, 4, 7:
                leftThumb = button[number]!
                ans += "L"

            case 3, 6, 9:
                rightThumb = button[number]!
                ans += "R"

            default:
                let leftThumbDist = leftThumb.manhattanDistance(button[number]!)
                let rightThumbDist = rightThumb.manhattanDistance(button[number]!)

                ans += leftThumbDist<rightThumbDist ? "L" : (leftThumbDist>rightThumbDist ? "R" : hand)
                leftThumb = ans.last! == "L" ? button[number]! : leftThumb
                rightThumb = ans.last! == "R" ? button[number]! : rightThumb
        }
    }
    
    return ans
}
